pydatastructs/linear_data_structures/arrays.py

```diff
@@ -133,13 +136,108 @@ def __len__(self):
     def __str__(self):
         return str(self._data)
 
+class MultiDimensionalArray(Array):
+    '''
+        Represents a multi-dimensional array.
+
+        Parameters
+        ==========
+
+        dtype: type
+            A valid object type.
+        size: int
+            The number of elements in the array.
+
+        Raises
+        ======
+        IndexError
+            Index goes out of boundaries
+        ValueError
+            When there's no dimensions or the dimension size is 0
+        TypeError
+            An argument is not what expected
+
+        Examples
+        ========
+        >>> from pydatastructs import MultiDimensionalArray as MDA
+        >>> arr = MDA(int, 5, 6, 9)
+        >>> arr.fill(32)
+        >>> arr[3][0][0]
+        32
+        >>> arr[3][0][0] = 7.2
+        >>> arr[3][0][0]
+        7
+
+        References
+        ==========
+
+        .. [1] https://en.wikipedia.org/wiki/Array_data_structure#Multidimensional_arrays
+        '''
+    __slots__ = ['_size', '_data', '_dtype']
+
+    def __new__(cls, dtype=NoneType, *args, **kwargs):
+        if dtype is NoneType or len(args) == (0):
+            raise ValueError("array cannot be created due to incorrect"
+                             " information.")
+        dimensions = list(args)
```

Comment:
What is this variable for?

Reply:
The variable is to store the sizes of the different dimensions, and used to set the size of the dimension
lets say you want to create an array of three dimensions, each dimension with its own size like 4 3 2
So, 
`MultiDimensionalArray(int, 4, 3, 5)`
then 
`dimensions == [4, 3, 5]`

I used it mostly for convenience since i use pop and I wanted to have an easier name, but if you want me to only use args, i can change it.

Reply:
Well, it appears to me that this uses recursive solution i.e., array of arrays of arrays for three dimensions. What we want is, an internal 1D layout with size as product of dimensions with RMO being used to convert queries like, M[i, j, k] to M[l], where l = RMO(i, j, k).

Reply:
ok, i did it again, this time it only uses a one dimensional array, calculating the position with the size of each dimension